java8/proxy/src/main/java/openwhisk/java/action/Proxy.java

```diff
@@ -14,156 +14,198 @@
-        byte[] bytes = content.getBytes(StandardCharsets.UTF_8);
-        t.sendResponseHeaders(code, bytes.length);
-        OutputStream os = t.getResponseBody();
-        os.write(bytes);
-        os.close();
+    Proxy.writeCodeOnFileSystem(vertx, jarBinary, ar -> {
+      if (ar.failed()) {
+        rc.fail(ar.cause());
+      } else {
+        if (!ar.result().toFile().isFile()) {
+          rc.response().setStatusCode(HTTP_INTERNAL_SERVER_ERROR).end("Error invoking function");
+        }
+        try {
+          final URL url = ar.result().toUri().toURL();
+          urlClassLoader = new URLClassLoader(new URL[] { url});
+          String clazzName = value.getString("main");
+          String methodName = "main";
+          if(clazzName == null || clazzName.isEmpty()){
+            clazzName = getClassNameFromManifest(url);
+          }
+          if(clazzName.indexOf('#')>-1){
+            final String[] splitted = clazzName.split("#");
+            clazzName = splitted[0];
+            methodName= splitted[1];
+          }
+          if(clazzName == null ){
+            rc.response().setStatusCode(ERROR_FUNCTION_NOT_FOUND).end("Main class is empty, can not determine the function to invoke");
+            return;
+          }
+          final Class<?> mainClass = urlClassLoader.loadClass(clazzName);
+
+          final Method m = mainClass.getMethod(methodName, new Class[] { com.google.gson.JsonObject.class });
+          m.setAccessible(true);
+          final int modifiers = m.getModifiers();
+          if (m.getReturnType() != com.google.gson.JsonObject.class || !Modifier.isStatic(modifiers) || !Modifier.isPublic(modifiers)) {
+            throw new NoSuchMethodException(methodName);
+          }
+          this.mainMethod = m;
+
+          rc.response().setStatusCode(HTTP_OK).end("OK");
+
+        }catch(NoSuchMethodException e){
+          rc.response().setStatusCode(ERROR_FUNCTION_NOT_FOUND).end(e);
+        }
+        catch (ClassNotFoundException | IOException e) {
+          rc.response().setStatusCode(HTTP_INTERNAL_SERVER_ERROR).end("An error has occurred (see logs for details): " + e);
+        }
+      }
+    });
+  }
+
+  private String getClassNameFromManifest(URL url ) throws IOException{
+    final URL jarUrl = new URL("jar", "", url + "!/");
+    final JarURLConnection connection = (JarURLConnection)jarUrl.openConnection();
+    final Attributes attributes = connection.getMainAttributes();
+    return attributes != null
+                   ? attributes.getValue(Attributes.Name.MAIN_CLASS)
+                   : null;
+  }
+
+  private void runHandler(RoutingContext rc) {
+    final JsonObject request = rc.getBodyAsJson();
+    final JsonObject value = request.getJsonObject("value");
+    final JsonParser parser = new JsonParser();
+    final com.google.gson.JsonObject req = parser.parse(value.toString()).getAsJsonObject();
```

Comment:
This code does:

1. parse body to a `io.vertx.core.json.JsonObject` (`rc.getBodyAsJson()`)
2. serialize `io.vertx.core.json.JsonObject` to a `String` (`value.toString()`)
3. parse the serialized value to a `com.google.gson.JsonObject` (`parser.parse()`)

Looks like unnecessary overhead.